src/core/border_router/nat64_translator.hpp

```diff
@@ -0,0 +1,185 @@
+     * be modified.
+     *
+     * @param[in,out] aMessage the message to be processed.
+     *
+     * @returns Result::kForward the caller should contiue forwarding the packet.
+     * @returns Result::kDrop the caller should drop the packet silently.
+     * @returns Result::kReplyIcmp the caller should reply an ICMP packet, the buffer will be filled with the content of
+     * the ICMP packet.
+     *
+     */
+    Result HandleOutgoing(Message &aMessage);
+
+    /**
+     * @brief This function sets the CIDR used when setting the source address of the outgoing translated IPv4 packets.
+     * A valid CIDR must have a non-zero prefix length.
+     *
+     * @note The actual addresses pool is limited by the size of the mapping pool and the number of addresses available
+     * in the CIDR block. If the provided is a valid IPv4 CIDR for NAT64, and it is different from the one already
+     * configured, the NAT64 translator will be reset and all existing sessions will be expired.
+     *
+     * @param[in] aCidr the CIDR for the sources of the translated packets.
+     *
+     * @retval  kErrorInvalidArgs    The the given CIDR a valid CIDR for NAT64.
+     * @retval  kErrorNone           Successfully enabled/disabled the NAT64 translator.
+     *
+     */
+    Error SetIp4Cidr(const Ip4::Cidr &aCidr);
+
+    /**
+     * @brief This function sets the prefix of NAT64-mapped addresses in the thread network. The address mapping table
+     * will not be cleared.
+     *
+     * @param[in] aNat64Prefix the prefix of the NAT64-mapped addresses.
+     *
+     */
+    void SetNat64Prefix(const Ip6::Prefix &aNat64Prefix);
+
+private:
+    class AddressMapping : public LinkedListEntry<AddressMapping>
+    {
+    public:
+        friend class LinkedListEntry<AddressMapping>;
+        friend class LinkedList<AddressMapping>;
+
+        void Touch(uint64_t aNow) { mExpiry = aNow + kAddressMappingIdleTimeoutMsec; }
+
+        Ip4::Address mIp4;
+        Ip6::Address mIp6;
+
+        // The timestamp when this mapping expires, in milliseconds.
+        uint64_t mExpiry;
+
+    private:
+        bool Matches(const Ip4::Address &aIp4) const { return mIp4 == aIp4; }
+        bool Matches(const Ip6::Address &aIp6) const { return mIp6 == aIp6; }
+        bool Matches(const uint64_t aNow) const { return mExpiry < aNow; }
+
+        AddressMapping *mNext;
+    };
+
+    Error TranslateIcmp4(AddressMapping &aMapping, Message &aMessage);
+
+    Error TranslateIcmp6(AddressMapping &aMapping, Message &aMessage);
+
+    void ReleaseMapping(AddressMapping &aMapping);
+
+    AddressMapping *CreateMapping(const Ip6::Address &aAddr);
+
+    AddressMapping *GetMapping(const Ip6::Address &aAddr, bool aTryCreate);
+
+    AddressMapping *GetMapping(const Ip4::Address &aAddr);
+
+    uint32_t     mAvailableAddressCount;
+    Ip4::Address mIp4AddressPool[kAddressMappingPoolSize];
```

Comment:
Can consider using `Array<Ip4::Address, kAddressMappingPoolSize>` for this? 
- It seems to be array of available IPv4 address that we can use.
- We pop entries from back of array and then place then back in the array when we release the mapping entry.
- So fits well with `Array` class and we can use `PopBack()` and `PushBack()` method of `Array().`

Reply:
Yes, Array is exactly what we want here. Thanks for notice.